### PR TITLE
Improve subsurface tests

### LINKED ISSUE #82: Should test subsurface reordering
These are currently not implemented in Mir, but non the less, there should be tests for them

### LINKED ISSUE #163: Clean up subsurface tests
Current subsurface tests have some issues (which I take full responsibility for):
- Long tests that test many things
- Bizarrely confusing assert messages
- ~Their own implementation of input and surface abstractions, which should be merged with the ones added in #156~ (this is now #171)

### PR DESCRIPTION
**Pull request by wmww**
- Split up long, convoluted tests
- Improve incomprehensible error messages
- Add subsurface reordering tests

Fixes #82, Fixes #163

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
> By my reading, all the `wl_subsurface.*` calls (with the exception of `set_sync` and `set_desync`) are not gated on the surface's `wl_surface.commit`, but instead on the _parent_ surface's `wl_surface.commit`, and can be updated without calling `wl_surface.commit` on the subsurface at all.

I don't follow how you arrive at this reading. My reading is that this behaviour is dependent on the synchronized/desynchronized mode:
- synchronized mode: the surface's wl_surface.commit "caches" the pending wl_surface state to be applied by the parent wl_surface.commit
- desynchronized mode: the surface's wl_surface.commit applies the pending wl_surface state directly

**Comment by wmww** [comment]
`set_sync`:
> In synchronized mode, wl_surface.commit on a sub-surface will
	accumulate the committed state in a cache, but the state will
	not be applied and hence will not change the compositor output.
	The cached state is applied to the sub-surface immediately after
	the parent surface's state is applied.

`set_desync`:
> In desynchronized mode, wl_surface.commit on a sub-surface will
	apply the pending state directly, without caching, as happens
	normally with a wl_surface.

**Comment by RAOF** [comment]
So, my reading is based on

> The behaviour of a wl_surface.commit request… [talks a bunch about how wl_surface state is handled].
> 
> Sub-surfaces have also other kind of state which is managed by wl_subsurface requests,… This [wl_*sub*surface] state is applied when the parent surface's wl_surface state is applied, regardless of the sub-surface's mode.

By my reading there are two distinct classes of state - wl_surface state, and wl_subsurface state - with different behaviour for when they get applied. The wl_surface state for surface `X` always requires a `wl_surface.commit` call on `X`, and depending on sync/desync mode may require some other set of `commit`s. The wl_subsurface state for surface `X` does *not* require a `wl_surface.commit` on surface `X`; rather, it is only dependent on when the parent of `X`'s wl_surface state is applied.

So, I *agree* with
>    synchronized mode: the surface's wl_surface.commit "caches" the pending wl_surface state to be applied by the parent wl_surface.commit
>    desynchronized mode: the surface's wl_surface.commit applies the pending wl_surface state directly

I just don't think that the wl_subsurface calls such as `wl_subsurface.place_above` *are* wl_surface state.

**Comment by RAOF** [comment]
From #wayland:

> RAOF: The position and stacking order of subsurfaces is part of the parent's state, and is applied on the parent's wl_surface.commit.

**Comment by wmww** [comment]
I updated tests to (hopefully) be consistent with the protocol. They now mostly fail on Mir because Mir was implemented based on my incorrect reading.

**Review by RAOF — APPROVED** [review]
Existing, but while we're here: `move_subsurface_to` should probably only commit the parent?

If you agree the issues raised are nits; please fix them and then bors.

bors delegate=wmww

**Comment by wmww** [comment]
Heads up I added a new test so I didn't bors it myself

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
I think the code is right, but by my reading there are three collections of state to be considered:

1. The `wl_surface` state of `main_surface`
2. The `wl_surface` state of `subsurface` (which may be sync or desync with 1, according to 3)
3. The `wl_subsurface` state of `subsurface` which is sync'ed with 1 not 2 (with the exception of set_sync()/set_desync())

Comments that say 3 is part of 1 conflate the rules for committing changes with the objects that hold the state.

File: tests/subsurfaces.cpp [hunk 1]
```diff
@@ -210,6 +209,7 @@ TEST_P(SubsurfaceTest, sync_subsurface_moves_when_only_parent_committed)
     int const subsurface_x = 20, subsurface_y = 20;
 
     wl_subsurface_set_position(subsurface, subsurface_x, subsurface_y);
+    // Position is part of parent (main_surface) state, so subsurface does not need to be committed
```
[the remark this answers is not in the document]
Comment (wmww) [reply tests/subsurfaces.cpp]: Perhaps fixed

**Review by AlanGriffiths — APPROVED** [review]
LGTM

bors r+

**Pull request merged**